### PR TITLE
MBS-10988 fix test artifacts access

### PR DESCRIPTION
**Pull request by dsvoronin**

### CODE REVIEW

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/report/listener/GsonReportParser.kt [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import java.io.File
+import java.io.FileReader
+
+internal class GsonReportParser(private val gson: Gson = reportGson) : ReportParser {
+
+    @Suppress("IfThenToElvis")
+    override fun parse(reportFile: File): Result<TestRuntimeData> {
+        return Result.tryCatch {
+            val testRuntimeData: TestRuntimeData? = FileReader(reportFile).use {
+                gson.fromJson(FileReader(reportFile), TestRuntimeDataPackage::class.java)
```
Comment (sergioboi) [inline subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/report/listener/GsonReportParser.kt]: FileReader(reportFile) twice

File: subprojects/common/report-viewer/src/main/kotlin/com/avito/report/TestArtifactsProviderFactory.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+            testDirGenerator = TestDirGenerator.Impl(className, methodName)
+        )
+    }
+
+    fun createForTempDir(tempDir: File): TestArtifactsProvider {
+        return DirectTestArtifactsProvider(lazy { tempDir })
+    }
+
+    @Suppress("SdCardPath") // android API's are unavailable here
+    fun createForAdbAccess(appUnderTestPackage: String, className: String, methodName: String): TestArtifactsProvider {
```
Comment (dsvoronin) [inline subprojects/common/report-viewer/src/main/kotlin/com/avito/report/TestArtifactsProviderFactory.kt:36]: what was wrong: i expected report files in wrong places

actual fix: access from different directories:

1. `ExternalStorageTransport` writes to `getExternalFilesDirs(targetContenxt)`
2. `DeviceWorker` specifies path on device via  `createForAdbAccess()` (to be more precise: `/sdcard/Android/data/${appUnderTestPackage}files`)
3. `ArtifactsTestListener` creates tempDirectory and pulls its content using path provided by `DeviceWorker`
4. `TestArtifactsProcessor` reads from tempDirectory and uploads multiple files in parallel but in blocking manner
5. `ArtifactsTestListener` removes tempDirectory after that

File: subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/report/listener/GsonReportParser.kt (hunk 1 above)
Comment (dsvoronin) [reply subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/internal/report/listener/GsonReportParser.kt]: fixed

File: subprojects/common/files/src/main/kotlin/com/avito/utils/Path.kt [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package com.avito.utils
+
+import com.avito.android.Result
+import java.nio.file.Path
+
+/**
+ * WARNING: only android 26+ supported
+ */
+@Suppress("NewApi")
```
Comment (eugene-krivobokov) [inline subprojects/common/files/src/main/kotlin/com/avito/utils/Path.kt:9]: Suppressing here hides the restrictions.
If we consume this module in an Android library, we wouldn't see the problem.
Let's use `@RequiresApi`.

File: subprojects/common/report-viewer/src/main/kotlin/com/avito/report/internal/DirectTestArtifactsProvider.kt [hunk 4]
```diff
@@ -0,0 +1,40 @@
+package com.avito.report.internal
+
+import com.avito.android.Result
+import com.avito.report.TestArtifactsProvider
+import java.io.File
+import java.util.UUID
+
+class DirectTestArtifactsProvider(
```
Comment (eugene-krivobokov) [inline subprojects/common/report-viewer/src/main/kotlin/com/avito/report/internal/DirectTestArtifactsProvider.kt:8]: Can it be internal as hidden by TestArtifactsProviderFactory?

File: subprojects/common/report-viewer/src/main/kotlin/com/avito/report/internal/UniqueDirTestArtifactsProvider.kt [hunk 5]
```diff
@@ -1,22 +1,17 @@
...
 import com.avito.android.Result
-import com.avito.report.ReportFileProvider
+import com.avito.report.TestArtifactsProvider
 import com.avito.report.TestDirGenerator
 import java.io.File
-import java.io.IOException
-import java.util.UUID
 
-class ReportFileProviderImpl(
+class UniqueDirTestArtifactsProvider(
```
Comment (eugene-krivobokov) [inline subprojects/common/report-viewer/src/main/kotlin/com/avito/report/internal/UniqueDirTestArtifactsProvider.kt:8]: https://github.com/avito-tech/avito-android/pull/938/files#r615603344

File: subprojects/gradle/runner/client/src/main/kotlin/com/avito/runner/scheduler/listener/ArtifactsTestListener.kt [hunk 6]
```diff
@@ -43,17 +45,24 @@ internal class ArtifactsTestListener(
...
-            is Failed.InRun ->
-                pullArtifacts(
-                    device = device,
-                    runnerOutputDirectory = testMetadataDirectory,
-                    testFolder = testFolder
-                )
+            is Failed.InRun -> {
+                val artifacts = testArtifactsDir.flatMap { dir ->
+
+                    // last /. means to adb to copy recursively, and not to copy last dir
```
Comment (eugene-krivobokov) [inline subprojects/gradle/runner/client/src/main/kotlin/com/avito/runner/scheduler/listener/ArtifactsTestListener.kt:57]:
```suggestion
                    // last /. means to adb to copy recursively, and do not copy the last dir
```

File: subprojects/gradle/runner/client/src/main/kotlin/com/avito/runner/scheduler/listener/ArtifactsTestListener.kt [hunk 7]
```diff
@@ -43,17 +45,24 @@ internal class ArtifactsTestListener(
...
+                    // example:
+                    //  - from: /sdcard/Android/someDir/ to: /xx ; will copy to /xx/someDir/ and not recursive
+                    //  - from: /sdcard/android/someDir/. to: /xx ; will copy to /xx and recursive
```
Comment (eugene-krivobokov) [inline subprojects/gradle/runner/client/src/main/kotlin/com/avito/runner/scheduler/listener/ArtifactsTestListener.kt:60]: Is it ADB specific? 
If so, let's keep this knowledge in `Device.pull` kdoc but not in clients.

File: subprojects/gradle/runner/client/src/main/kotlin/com/avito/runner/scheduler/listener/ArtifactsTestListener.kt [hunk 8]
```diff
@@ -43,17 +45,24 @@ internal class ArtifactsTestListener(
...
+                    // todo move this knowledge under adb layer
```
Comment (eugene-krivobokov) [inline subprojects/gradle/runner/client/src/main/kotlin/com/avito/runner/scheduler/listener/ArtifactsTestListener.kt:61]: Why postpone it if it's just a comment?

**Review by eugene-krivobokov — APPROVED** [review]

File: subprojects/common/files/src/main/kotlin/com/avito/utils/Path.kt (hunk 3 above)
Comment (dsvoronin) [reply subprojects/common/files/src/main/kotlin/com/avito/utils/Path.kt:9]: can't without adding android annotations; do you want me to add it?

File: subprojects/common/report-viewer/src/main/kotlin/com/avito/report/internal/DirectTestArtifactsProvider.kt (hunk 4 above)
Comment (dsvoronin) [reply subprojects/common/report-viewer/src/main/kotlin/com/avito/report/internal/DirectTestArtifactsProvider.kt:8]: yes

File: subprojects/gradle/runner/client/src/main/kotlin/com/avito/runner/scheduler/listener/ArtifactsTestListener.kt (hunk 8 above)
Comment (dsvoronin) [reply subprojects/gradle/runner/client/src/main/kotlin/com/avito/runner/scheduler/listener/ArtifactsTestListener.kt:61]: it's urgent, blocking release

File: subprojects/common/files/src/main/kotlin/com/avito/utils/Path.kt (hunk 3 above)
Comment (eugene-krivobokov) [reply subprojects/common/files/src/main/kotlin/com/avito/utils/Path.kt:9]: Yes, why not?
We don't forbid the usage of the common libraries in Android libraries.

Comment (dsvoronin) [reply subprojects/common/files/src/main/kotlin/com/avito/utils/Path.kt:9]: yep, thanks

File: subprojects/common/report-viewer/src/main/kotlin/com/avito/report/TestArtifactsProviderFactory.kt [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        className: String,
+        methodName: String
+    ): TestArtifactsProvider {
+        return UniqueDirTestArtifactsProvider(
+            rootDir = testReportRootDir,
+            testDirGenerator = TestDirGenerator.Impl(className, methodName)
+        )
+    }
+
+    fun createForTempDir(tempDir: File): TestArtifactsProvider {
```
Comment (eugene-krivobokov) [inline subprojects/common/report-viewer/src/main/kotlin/com/avito/report/TestArtifactsProviderFactory.kt:31]: Can and should we know about "temp" nature of the dir at this level?
Maybe we could express other attributes of dir here if they are important.

**Pull request merged**